pyteal/compiler/compiler.py

```diff
@@ -133,26 +133,62 @@ def compileSubroutine(
             ast = Return(ast)
 
     options.setSubroutine(currentSubroutine)
+
     start, end = ast.__teal__(options)
     start.addIncoming()
     start.validateTree()
 
-    start = TealBlock.NormalizeBlocks(start)
-    start.validateTree()
+    if (
+        currentSubroutine is not None
+        and currentSubroutine.get_declaration().deferred_expr is not None
+    ):
+        deferred_expr = cast(Expr, currentSubroutine.get_declaration().deferred_expr)
+        # these blocks represent a path of code that should be inserted before each retsub op
+        get_deferred_blocks = lambda: deferred_expr.__teal__(options)
+
+        # insert deferred blocks where needed
+        for block in TealBlock.Iterate(start):
+            if not isinstance(block, TealSimpleBlock):
+                continue
+
+            if not any(op.getOp() == Op.retsub for op in block.ops):
+                continue
+
+            if len(block.ops) != 1:
+                # we expect all retsub ops to be in their own block at this point
+                raise TealInternalError(
+                    f"Expected retsub to be the only op in the block, but there are {len(block.ops)} ops"
+                )
 
-    order = sortBlocks(start, end)
-    teal = flattenBlocks(order)
+            # insert deferred blocks between the previous block(s) and this one
+            deferred_start, deferred_end = get_deferred_blocks()
+            deferred_start.addIncoming()
+            deferred_start.validateTree()
 
-    verifyOpsForVersion(teal, options.version)
-    verifyOpsForMode(teal, options.mode)
+            deferred_start.incoming = block.incoming
+            block.incoming = [deferred_end]
+            deferred_end.nextBlock = block
+
+            for prev in deferred_start.incoming:
+                prev.replaceOutgoing(block, deferred_start)
+
+            if block is start:
+                # this is the start block, replace start
+                start = deferred_start
```

Comment:
For reviewers studying this code, it's very similar to the implementation of `TealBlock.NormalizeBlocks`: https://github.com/algorand/pyteal/blob/5596a71c5de238a33167cc5cb940f7397225d0d9/pyteal/ir/tealblock.py#L187